### PR TITLE
Update best practices for entrypoint.

### PR DESCRIPTION
**Pull request by SvenDowideit**
takeover #10470

@jfrazelle :)

### CODE REVIEW

**Comment by SvenDowideit** [comment]
@fredlf @jamtur01

File: docs/sources/articles/dockerfile_best-practices.md [hunk 1]
```diff
@@ -291,32 +291,28 @@ auto-extraction capability, you should always use `COPY`.
 
 ### [`ENTRYPOINT`](https://docs.docker.com/reference/builder/#entrypoint)
 
-The best use for `ENTRYPOINT` is as a helper script. Using `ENTRYPOINT` for
-other tasks can make your code harder to understand. For example,
+The best use for `ENTRYPOINT` is to set the image's main command, allowing that
+image to be run as though it was that command (and then to use `CMD` the default
```
Comment (duglin) [inline docs/sources/articles/dockerfile_best-practices.md]: I think its missing an "as".
...then to use `CMD` *as* the default options).

File: docs/sources/articles/dockerfile_best-practices.md [hunk 2]
```diff
@@ -291,32 +291,28 @@ auto-extraction capability, you should always use `COPY`.
...
 
-....docker run -it --entrypoint bash official-image -i
+Now the image can be run like this:
 
-This is especially true for new Docker users, who might naturally assume the
-above command will work fine. In cases where an image uses `ENTRYPOINT` for
-anything other than just a wrapper script, the command will fail and the
-beginning user will then be forced to learn about `ENTRYPOINT` and
-`--entrypoint`.
+    $ docker run s3cmd ls s3://mybucket
```
Comment (duglin) [inline docs/sources/articles/dockerfile_best-practices.md]: It might be good to show running the image twice, once with the "ls" parameters and once without them. That way it shows how/when `--help` will be used.  Also, perhaps show the output (or explain what the output would be) in each case.

File: docs/sources/articles/dockerfile_best-practices.md [hunk 3]
```diff
@@ -335,12 +331,31 @@ fi
 exec "$@"
 ```
 
-That script then gets copied into the container and run via `ENTRYPOINT` on
-container startup:
+> **Note**: This script uses [the `exec` Bash command](http://wiki.bash-hackers.org/commands/builtin/exec)
+> so that the final running application becomes the container's `PID 1` so that
+> it will receive any Unix signals sent to the container. See the
```
Comment (thaJeztah) [inline docs/sources/articles/dockerfile_best-practices.md]: I don't like the "so that" *twice* in this sentence. Also "This script" and "it will receive" may need to be more specific, it's a bit hard to follow what they refer to.

File: docs/sources/articles/dockerfile_best-practices.md [hunk 4]
```diff
@@ -335,12 +331,31 @@ fi
...
+> [`ENTRYPOINT`](https://docs.docker.com/reference/builder/#ENTRYPOINT) help for more
+> details.
+
+
+That script is copied into the container and run via `ENTRYPOINT` on
```
Comment (thaJeztah) [inline docs/sources/articles/dockerfile_best-practices.md]: "That script" possibly confusing again (I "get" the difficulty here :))

"... and run via"; should that be "... and *executed* via (/by?)" ? Not sure, may be just preference.

File: docs/sources/articles/dockerfile_best-practices.md [hunk 5]
```diff
@@ -335,12 +331,31 @@ fi
...
+This script will allow the user to just start Postgres:
+
+    $ docker run postgres
+
+or, add some parameters to the Postgres server:
+
+    $ docker run postgres postres --help
+
+or, to start a totally different tool, such as Bash:
+
```
Comment (thaJeztah) [inline docs/sources/articles/dockerfile_best-practices.md]: "or, to start" -> "or, start"?

The repeated "or, " overall reads a bit funny. Perhaps;

This script (will) allow(s) the user to just start Postgres (..) add some parameters .. (..) or start a different tool...

File: docs/sources/articles/dockerfile_best-practices.md (hunk 3 above)
Comment (fredlf) [inline docs/sources/articles/dockerfile_best-practices.md]: Mmm hmm, I think there is more stress-worthy material in this one sentence than there are stress positions available so.... let's break it into two sentences:
"This script uses [the `exec` Bash command](http://wiki.bashhackers.org/commands/builtin/exec)
so that the final running application becomes the container's `PID 1`. This allows the application to receive any Unix signals sent to the container."

**Comment by jfrazelle** [comment]
yayyy thanks @SvenDowideit you da best :D

File: docs/sources/articles/dockerfile_best-practices.md (hunk 4 above)
Comment (fredlf) [inline docs/sources/articles/dockerfile_best-practices.md]: I think the passive voice is the issue making things unclear. What is causing the script to be copied and run? Need to make the agent of that action clearer. Some more context cues might help too, like explicitly referring to the "helper script" (not just "script").

File: docs/sources/articles/dockerfile_best-practices.md [hunk 6]
```diff
@@ -335,12 +331,31 @@ fi
...
+> details.
+
+
+That script is copied into the container and run via `ENTRYPOINT` on
+container start:
 
     COPY ./docker-entrypoint.sh /
     ENTRYPOINT ["/docker-entrypoint.sh"]
 
+This script will allow the user to just start Postgres:
```
Comment (fredlf) [inline docs/sources/articles/dockerfile_best-practices.md]: It will help to give the reader a heads up for what is coming. E.g., "This script allows the user to interact with Postgres in several ways. It can simply start Postgres:  `$ docker run postgres`. Or, it can be used to run Postgres and pass parameters to the server: `$ docker run postgres postres --help`. Lastly, it could also be used to start a totally different tool, such Bash:

**Comment by fredlf** [comment]
Needs a few edits/revisions to clear up some ambiguities.

**Comment by SvenDowideit** [comment]
@fredlf @jamtur01 @spf13 @moxiegirl

File: docs/sources/articles/dockerfile_best-practices.md [hunk 7]
```diff
@@ -291,32 +291,32 @@ auto-extraction capability, you should always use `COPY`.
 
 ### [`ENTRYPOINT`](https://docs.docker.com/reference/builder/#entrypoint)
 
-The best use for `ENTRYPOINT` is as a helper script. Using `ENTRYPOINT` for
-other tasks can make your code harder to understand. For example,
+The best use for `ENTRYPOINT` is to set the image's main command, allowing that
+image to be run as though it was that command (and then use `CMD` as the default
+options).
```
Comment (jamtur01) [inline docs/sources/articles/dockerfile_best-practices.md]: flags?

File: docs/sources/articles/dockerfile_best-practices.md [hunk 8]
```diff
@@ -291,32 +291,32 @@ auto-extraction capability, you should always use `COPY`.
...
 
-For example, let’s look at the `Dockerfile` for the
-[Postgres Official Image](https://github.com/docker-library/postgres).
-It refers to the following script: 
+    $ docker run s3cmd ls s3://mybucket
+
+This is useful because the image name can double as a reference to the binary as
+shown in the command above.
+
+The `ENTRYPOINT` can also be used in combination with a helper script, allowing
```
Comment (jamtur01) [inline docs/sources/articles/dockerfile_best-practices.md]: instruction?

File: docs/sources/reference/builder.md [hunk 9]
```diff
@@ -625,9 +625,7 @@ ENTRYPOINT ["/usr/sbin/apache2ctl", "-D", "FOREGROUND"]
 ```
 
 If you need to write a starter script for a single executable, you can ensure that
-the final executable receives the Unix signals by using `exec` and `gosu`
-(see [the Dockerfile best practices](/articles/dockerfile_best-practices/#entrypoint)
-for more details):
+the final executable receives the Unix signals by using `exec` and `gosu`:
```
Comment (jamtur01) [inline docs/sources/reference/builder.md]: by using the ... commands?

**Comment by SvenDowideit** [comment]
@fredlf @jamtur01 @spf13 @moxiegirl

**Comment by jamtur01** [comment]
LGTM

File: docs/sources/reference/builder.md [hunk 10]
```diff
@@ -626,8 +626,7 @@ ENTRYPOINT ["/usr/sbin/apache2ctl", "-D", "FOREGROUND"]
 
 If you need to write a starter script for a single executable, you can ensure that
 the final executable receives the Unix signals by using `exec` and `gosu`
-(see [the Dockerfile best practices](/articles/dockerfile_best-practices/#entrypoint)
-for more details):
```
Comment (fredlf) [inline docs/sources/reference/builder.md]: Why did we remove this link?

Comment (SvenDowideit) [inline docs/sources/reference/builder.md]: Because that link doesn't tell the reader anything about `exec` and `gosu` - and much of the content about using `ENTRYPOINT` isn't there, its in the builder doc.

Comment (fredlf) [inline docs/sources/reference/builder.md]: Oh, okay. I see there's a link to the Best Practices doc at the top of the page, so I'm fine removing this one, then.

**Comment by fredlf** [comment]
LGTM
